features/332-sso/api.md

```diff
@@ -0,0 +1,311 @@
+
+## Possible Vulnerability
+
+Suppose Alice grants access to Mallory, a malicious application, via OAuth 2.0. Mallory now has the bearer access token of Alice, which is considered a proof of identity. Mallory can now impersonate Alice and login our application with the identity of Alice.
+
+Since this API is vulnerable by design, it is disabled by default. The developer is strongly recommended to use the standard OAuth 2.0 Authorization Code flow even the application is native and has already had acquired an access token.
+
+### API
+
+```typescript
+function loginOAuthProviderWithAccessToken(providerID: string, accessToken: string, options?: LoginOptions): Promise<User>;
+function linkOAuthProviderWithAccessToken(providerID: string, accessToken: string): Promise<User>;
+```
+
+### Example
+
+```typescript
+getAccessTokenViaFacebookSDKSomehow().then(accessToken => {
+  loginOAuthProviderWithAccessToken("facebook", accessToken).then(user => {
+    // login suceedeed
+  });
+});
+```
+
+# Remove OAuth Provider
+
+If the removal will result in no login principals, then removal is forbidden. Such case happens when the user logged in with Google and has no other login means.
+
+### API
+
+```typescript
+function removeOAuthProvider(providerID: string): Promise<void>;
+```
+
+### Example
+
+```typescript
+askConfirmation().then(confirmed => {
+  if (confirmed) {
+    skygear.auth.removeOAuthProvider("google");
+  }
+});
+```
+
+# Merging existing user for login
+
+Some users may forget that they have signed up with email long ago. When they return to the application again, they may login with OAuth 2.0, creating a new user identity unexpectedly. To mitigate this issue, the developer can instruct the Auth gear on how to merge existing user.
+
+All login APIs accept `LoginOptions`. By default, merging is enabled. If `mergeRealm` is not in `allowedRealms`, it is an error and the login process is aborted. Otherwise, the implementation will merge existing user by inferring email from the provider user info and use it and the realm to match existing user.
+
+> In addition to email, should we consider phone number?
+
+# Removal of v1 API
+
+The following v1 API is removed.
+
+```typescript
+function getOAuthProviderProfiles(): Promise<{[providerID: string]: any}>;
+```
+
+The existing implementation simply return all snapshots of user info of all providers of the logged in user. The snapshot of user info most likely will be stale. So the function unlikely return what the developer expects. If we want to make the function always return the fresh user info, we need to implement refresh token of OAuth 2.0. However, to implement refresh token, we have to drop support for `withAccessToken`, not to mention that some providers like Facebook does not support refresh token.
```

Comment:
So we don't handle refresh of data first right? (Is ok, I don't think we handled it before neither actually)

Reply:
Yes, v1 did not implement refresh token. So I am proposing to remove this confusing API first. We can add it back when refresh token is implemented.